add "replace" and "upload" functionality

liyakun opened the pull request:
![replace1](https://cloud.githubusercontent.com/assets/7370334/7135065/4fb2a15e-e2a3-11e4-9d16-436263a39c4d.png)
![replace2](https://cloud.githubusercontent.com/assets/7370334/7135062/4fb0ae08-e2a3-11e4-90f6-e3090f58ebca.png)
![upload1](https://cloud.githubusercontent.com/assets/7370334/7135063/4fb0e436-e2a3-11e4-986c-984e1d5c3b44.png)
![upload2](https://cloud.githubusercontent.com/assets/7370334/7135064/4fb0e6ac-e2a3-11e4-8ca6-7353e2b400cb.png)

add "replace" and "upload" functionalities

add "upload" file name check

add upload service with commit_message_replace

add replace action

add "replace" view

show "replace"

add "replace" commit message

add "replace" and "upload" test

add "replace" and "upload" test

support "replace" and "upload" functionalities

add "upload" view

add upload button

show "upload"

new functionality added

randx commented:
Please fix next: 

* no tabs - only spaces
* no javascript but coffescript
* CHANGELOG item looks out of version block
* I dont like this 2 buttons: `Create`and `Upload`. Creating of file should be always one link to page where you either write content or upload it

liyakun commented:
@randx 

Hi,
I already changed the code according to your last two comments. I am not clear what you first two comments mean.

1. no tabs - only spaces.
Do you mean in the code?

2. no javascript but coffescript
I use javascript some part, like checking the file uploaded or not. This is because the 
file used in original project, like _remove.htmal.haml also use javascript. 

If my understanding is wrong, please correct me. Thank you very much.

liyakun commented:
The view for "upload" is like this now

![download3](https://cloud.githubusercontent.com/assets/7370334/7201385/a9818f50-e506-11e4-9dca-abfc505c8b08.png)
![download4](https://cloud.githubusercontent.com/assets/7370334/7201386/a984e146-e506-11e4-9134-1e2b573205f2.png)
![download5](https://cloud.githubusercontent.com/assets/7370334/7201384/a9819446-e506-11e4-8ab7-e3536ddfb004.png)

randx commented:
@liyakun I propose we have next title `Create new file or upload existing one`. Then you can make ` upload existing one` as a link that open popup. What do you think?

randx commented:
> no tabs - only spaces.
Do you mean in the code?

yes

> no javascript but coffescript

Instead of 

```
:javascript 
  function checkFile(){
```

use 

```
:coffescript
  checkFile ->
```

liyakun commented:
@randx 
You mean the outlook is like this
![upload_existing](https://cloud.githubusercontent.com/assets/7370334/7233279/a5caab2c-e780-11e4-926a-8c2eb12bcd33.png)
![upload_existing1](https://cloud.githubusercontent.com/assets/7370334/7233280/a5ccf5e4-e780-11e4-8045-d96c2dc65edd.png)

randx commented:
@liyakun yes I think it looks nice in this way

liyakun commented:
@randx 
if using coffescript, then which coffeescript haml filter you prefer to use?

randx commented:
@liyakun sorry but I dont understand your question.

liyakun commented:
@randx hi,
I change the code to coffescript, but there is an error "Filter " 'coffescript " is not defined'.
So, I think should I use some coffeescript haml filter, then I ask which filter you recommend?

randx commented:
@liyakun here is example - https://github.com/gitlabhq/gitlabhq/blob/master/app/views/users/show.html.haml#L49-L50

liyakun commented:
@randx I am not sure why this error always show here, but there is no this error in my local machine test. Thanks.

randx commented:
@liyakun we cant accept PR with broken tests

randx commented on CHANGELOG:
```diff
@@ -37,6 +37,14 @@ v 7.10.0
+v 7.10.0 (unreleased)
+  -Add "replace" and "upload" functionalities
+  -"replace" function allows user replace existing file with new content, and same content checking included.
+  -"upload" function allows user upload new file into current repository, and sane name file checking included.
```
FIx this

randx commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -21,14 +21,27 @@ def new
+        @ref,
+        file_path
+      ).execute
+    end
```
Refactor this

randx commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -48,18 +61,29 @@ def edit
+        @ref, 
+        @path
+    ).execute
+    end
```
should be refactored

randx commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,32 @@
+      alert('Please select file!');
+    }
+  }
+  disableButtonIfAnyEmptyField('#commit_message_replace')
```
should be coffeescript

randx replied:
and refactored

randx commented:
@liyakun This is 5th time I return to your PR and it still not addressed all comments. Please finish your PR, get `Ready for Merge` label from core team and after it ping me for review

liyakun commented:
@randx I changed the file according to your comment, however, I am not clear exactly what you expect from the "should be refactored". Could you give more hints on this part? I will do rebase after finishing this part. Thanks.

randx commented:
@liyakun you need to address DRY principle in this PR. I see a lot of duplicating code here

liyakun commented:
@randx thanks for your advice, just want to make sure you notice that the code is not using the same services. The if and else statements using different services, like "if Files::ReplaceService else Files::UpdateService", and also the parameters are not the same value.

73 commented:
Hi everyone,
we would like to see that functionality in gitlab. Since I can't make out a exact status for this PR from the comments above, I would like to ask whats the next thing thats suppose to happen here? What developer action/feedback is requested? (Since the last feedback was provided by the developer, I guess)

Thanks,
73

randx commented on CHANGELOG:
```diff
@@ -85,6 +88,11 @@ v 7.11.0
+  - Fix Error 500 when searching Wiki pages (Stan Hu)
+  - Add "replace" and "upload" functionalities
+  - "replace" function allows user replace existing file with new content, and same content checking included.
+  - "upload" function allows user upload new file into current repository, and sane name file checking included.
```
Please remove duplication from CHANGELOG

randx commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -146,10 +146,19 @@ def editor_variables
         @path
       end
+     
+    if !params[:file_upload].nil?
```
Replace `!variable.nil?` with `variable.present?`

randx commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -146,10 +146,19 @@ def editor_variables
 
     @file_path =
       if action_name.to_s == 'create'
+        if !params[:file_upload].nil?
```
Replace `!variable.nil?` with `variable.present?`

randx commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -42,7 +42,7 @@ def edit
   def update
     result = Files::UpdateService.new(@project, current_user, @commit_params).execute
-
+       
```
Do we need this spaces?

randx commented on app/views/projects/blob/new.html.haml:
```diff
@@ -16,6 +22,6 @@
     = render 'projects/commit_button', ref: @ref,
               cancel_path: namespace_project_tree_path(@project.namespace, @project, @id)
-
+              
```
Do we need this spaces?

randx commented on app/views/projects/blob/new.html.haml:
```diff
@@ -1,7 +1,13 @@
+  
+.file-title
+  =render 'projects/blob/upload'
+  </br>
```
Why html tag is here?

liyakun replied:
@randx , here the tag is used for get the upload modal, I tried other ways, however, this is the one succeed. Which way do you think should  be here?

randx replied:
`%br`

randx commented on app/views/projects/tree/_tree.html.haml:
```diff
@@ -12,7 +12,7 @@
       = link_to namespace_project_new_blob_path(@project.namespace, @project, @id), title: 'New file', id: 'new-file-link' do
         %small
-          %i.fa.fa-plus
+          %i.fa.fa-plus 
```
Do we need extra space here?

randx commented on features/steps/project/source/browse_files.rb:
```diff
@@ -106,8 +122,35 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
+  step 'I check name of the uploaded file' do
+    ".gitignore" != "user.feature"
+    "LICENSE" != "user.feature"
+    "VERSION" != "user.feature"
```
I dont understand how this works

liyakun replied:
@randx , as the new feature needs to test the upload file, so I use one of the project file "user.feature" for test upload. The upload function also needs to avoid the case same name file exists in the current repository, so I add "I check name of the uploaded file" which just means the name of file "user.feature" is different with the file in current repository.

randx replied:
You compare one string to different one, right? It will always be true - so I dont understand what behaviour is tested here

liyakun replied:
@randx ,  actually here is just a logic to say that the uploaded file "user.feature" has a different name with the file in current repository.

JobV replied:
@liyakun you are comparing two strings. The result of the comparison will _always_ be true, because these are not variables, just plain strings.

@randx is wondering why you would add this, as it doesn't help the tests. This part will always pass. It's the same as comparing `"1" != "0"`.

Did you made a mistake here or are you trying to do something else? Please explain what the purpose is.

liyakun replied:
@JobV  @randx here is for the scenario that when user upload a file, like ".gitignore", which has a same name with file in the current repository, this should be notified. 

As in this test, the uploaded file is "user.feature", it will will always pass. 

How do you think?  Thanks.

JobV replied:
@liyakun when writing a test, first try to write a failing test. If the test can't fail, why would you write it? 

Maybe you can make "user.feature" or ".gitignore" a variable, rather than a string? Otherwise we're not testing functionality, integration or any code. We are just adding a test that always passes. There is no value in that.

So to repeat, try writing a test that fails something and then write the code that has to pass it. In this case, you could comment out or remove part of your code so that this test will fail. Then you write the code to make it pass and refactor it.

Here is a nice website on it: http://www.jamesshore.com/Blog/Red-Green-Refactor.html

liyakun replied:
@JobV , thanks, I will change the test.

randx commented:
@73 thank you for reminder

@liyakun I left few comment for this PR

liyakun commented:
@73 thanks for your reminder

liyakun commented:
@randx I deleted the duplication  in CHANGELOG file, and the extra spaces in other files you have pointed out. I also added replay to your other comments, thanks.

randx commented on features/steps/project/source/browse_files.rb:
```diff
@@ -106,8 +122,35 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
+  end
+
+  step 'I click on "upload existing one"' do
+      click_link 'upload existing one'
```
fix indentation please

liyakun replied:
Hi, @randx  the indentation here fixed, thanks.

tsigo commented on app/views/projects/blob/_actions.html.haml:
```diff
@@ -17,6 +17,10 @@
+  = button_tag class: 'btn btn-small btn-primary btn-xs',
+      'data-toggle' => 'modal', 'data-target' => '#modal-replace-blob' do
+    Replace	 
+  = |
```
I don't think this line is necessary, is it?

liyakun replied:
@tsigo "= |" add a space between the two button

stanhu replied:
Could this be done with a button group?

http://getbootstrap.com/components/#btn-groups

liyakun replied:
@stanhu @tsigo , It can be done by button group, and I attach one screen shot, and will update the code to the repository, thanks.
![chinesereminder1](https://cloud.githubusercontent.com/assets/7370334/8374439/a0c99e18-1bf5-11e5-89bb-41d8797b7737.png)

tsigo commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,26 @@
+#modal-replace-blob.modal.hide
```
Remove the `.hide` class, it's no longer necessary.

tsigo commented on app/views/projects/blob/_upload.html.haml:
```diff
@@ -0,0 +1,26 @@
+#modal-upload-blob.modal.hide
```
Remove the `.hide` class, it's no longer necessary.

tsigo commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,26 @@
+        = form_tag namespace_project_blob_path(@project.namespace, @project, @id), method: :put, class: 'form-horizontal', :multipart => true  do
+          %br
+          .form-group
+            .col-sm-2
```
Remove the empty `.col-sm-2` class and use `.col-sm-offset-2` on the other div.

See https://github.com/gitlabhq/gitlabhq/commit/89bcc1baf040e998730fa7c3e029daf9112321eb

tsigo commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,26 @@
+          = render 'shared/commit_message_replace_container', params: params,
+                  placeholder: 'Replace this file because...'
+          .form-group
+            .col-sm-2
```
Remove the empty `.col-sm-2` class and use `.col-sm-offset-2` on the other div.

See https://github.com/gitlabhq/gitlabhq/commit/89bcc1baf040e998730fa7c3e029daf9112321eb

tsigo commented on app/views/projects/blob/_upload.html.haml:
```diff
@@ -0,0 +1,26 @@
+        = form_tag namespace_project_create_blob_path(@project.namespace, @project, @id), method: :post, class: 'form-horizontal', :multipart => true do
+          %br
+          .form-group
+            .col-sm-2
```
Remove the empty `.col-sm-2` class and use `.col-sm-offset-2` on the other div.

See https://github.com/gitlabhq/gitlabhq/commit/89bcc1baf040e998730fa7c3e029daf9112321eb

tsigo commented on app/views/projects/blob/_upload.html.haml:
```diff
@@ -0,0 +1,26 @@
+          = render 'shared/commit_message_replace_container', params: params,
+                  placeholder: 'Upload this file because...'
+          .form-group
+            .col-sm-2
```
Remove the empty `.col-sm-2` class and use `.col-sm-offset-2` on the other div.

See https://github.com/gitlabhq/gitlabhq/commit/89bcc1baf040e998730fa7c3e029daf9112321eb

tsigo commented on app/views/projects/blob/new.html.haml:
```diff
@@ -1,5 +1,11 @@
-%h3.page-title New file
+%h3.page-title<
+  Create new file or 
+  %a.button{:href=>'#modal-upload-blob', 'data-target'=>'#modal-upload-blob', 'data-toggle'=>'modal'}> &nbspupload existing one
```
`href` should use Ruby 1.9 Hash syntax.

You're also missing the semicolon for the `&nbsp;` entity.

tsigo commented on app/views/projects/blob/new.html.haml:
```diff
@@ -1,5 +1,11 @@
+  %a.button{:href=>'#modal-upload-blob', 'data-target'=>'#modal-upload-blob', 'data-toggle'=>'modal'}> &nbspupload existing one
+  
+.file-title
+  =render 'projects/blob/upload'
```
Space after `=`

tsigo commented on app/views/projects/blob/new.html.haml:
```diff
@@ -16,6 +22,6 @@
     = render 'projects/commit_button', ref: @ref,
               cancel_path: namespace_project_tree_path(@project.namespace, @project, @id)
-
+ 
```
Whitespace.

tsigo commented on app/views/shared/_commit_message_replace_container.html.haml:
```diff
@@ -0,0 +1,14 @@
+.form-group.commit_message-group
+  = label_tag 'commit_message_replace', class: 'control-label' do
```
Make the first argument to `label_tag` a symbol, just for consistency with the rest of the app.

liyakun replied:
@tsigo , you mean change the naming rule for 'commit_message_replace' ?

tsigo replied:
@liyakun No I mean just make the first argument a symbol.

`= label_tag :commit_message_replace, class: 'control-label' do`

liyakun replied:
@tsigo I checked another file _commit_message_container.html.haml in the newest repository, it also keep using with '**', so I keep the same, should I change?

tsigo replied:
@liyakun Yes, see my original note. That file is wrong too, but that doesn't mean yours has to be. :smile:

liyakun replied:
@tsigo okay, thanks

tsigo commented on app/views/shared/_commit_message_replace_container.html.haml:
```diff
@@ -0,0 +1,14 @@
+  .col-sm-10
+    .commit-message-replace-container
+      .max-width-marker
+      = text_area_tag 'commit_message_replace',
```
`:commit_message_replace`

liyakun replied:
@tsigo you mean change text_area_tag to :commit_message_replace ?

tsigo replied:
Similar to above.

`= text_area_tag :commit_message_replace,`

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -36,6 +40,10 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
   end
 
+  step 'I should see new file content' do
+    old_gitignore_content != '*.rbc'
```
This isn't doing anything.

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -20,6 +20,10 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
   end
 
+  step 'I see the "user.feature"' do
+    namespace_project_tree_path(@project.namespace, @project, 'master/user.feature') == true
```
This isn't doing anything.

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -106,8 +122,35 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
-    expect(page).to have_content "New file"
-    expect(page).to have_content "Commit message"
+    page.should have_content "new file"
+    page.should have_content "Commit message"
```
Update these for RSpec 3 syntax -- `expect(page).to have_content ...`

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -106,8 +122,35 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
+  
+  step 'I replace it with "LICENSE"' do
+    attach_file(:file_upload, "LICENSE")
+    old_gitignore_content = "LICENSE"
```
Why is this variable assigned? It's not an instance variable so it can't be used anywhere outside this step, and the step itself doesn't use it.

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -147,6 +190,10 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
   end
+  
+  step 'I am redirected to the "user.feature"' do
+    current_path.should == namespace_project_tree_path(@project.namespace, @project, 'master/user.feature')
```
RSpec 3 syntax.

liyakun commented:
@JobV @tsigo the test part was changed, and I removed the part not essentially works, thanks.

tsigo commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,24 @@
+        = form_tag namespace_project_blob_path(@project.namespace, @project, @id), method: :put, class: 'form-horizontal', :multipart => true  do
+          %br
+          .form-group
+            .col-sm-offset.2.col-sm-10
```
This line is wrong. It should be `.col-sm-offset-2.col-sm-10` (note the dash before the 2).

tsigo commented on app/views/projects/blob/_upload.html.haml:
```diff
@@ -0,0 +1,24 @@
+        = form_tag namespace_project_create_blob_path(@project.namespace, @project, @id), method: :post, class: 'form-horizontal', :multipart => true do
+          %br
+          .form-group
+            .col-sm-offset.2.col-sm-10
```
This line is wrong. It should be `.col-sm-offset-2.col-sm-10` (note the dash before the 2).

tsigo commented on features/project/source/browse_files.feature:
```diff
@@ -147,7 +167,7 @@ Feature: Project Source Browse Files
     Then I see Browse code link
 
-  # Permalink
+  # Permalink 
```
Trailing whitespace.

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -19,7 +19,7 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
     expect(page).to have_content '.gitignore'
   end
-
+  
```
Whitespace here and on line 38.

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -106,9 +114,29 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
   end
+  
+  step 'I can see "upload existing one"' do
+    page.should have_content "upload existing one"
```
RSpec 3 syntax please.

tsigo commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -42,7 +42,7 @@ def edit
   def update
     result = Files::UpdateService.new(@project, current_user, @commit_params).execute
-
+ 
```
Whitespace.

liyakun commented:
@tsigo , thank you very much, the commented parts are fixed.

tsigo commented on features/steps/project/source/browse_files.rb:
```diff
@@ -35,7 +35,7 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
     expect(page).to have_content new_gitignore_content
   end
-
+ 
```
Still whitespace here (sorry, I'm a stickler :smile:)

liyakun replied:
@tsigo , thanks, I need to do that, just in my local vim editor, I did not find whitespace there, sorry.

stanhu commented on CHANGELOG:
```diff
@@ -85,6 +86,8 @@ v 7.11.0
   - Don't allow a merge request to be merged when its title starts with "WIP".
   - Add a page title to every page.
   - Allow primary email to be set to an email that you've already added.
+  - Fix Error 500 when searching Wiki pages (Stan Hu)
```
I think these extraneous entries should not be here.

tsigo replied:
Yes I meant to check on that. It should just be your addition, @liyakun

stanhu replied:
Right, and this feature should be bumped up to v7.13.0.

stanhu commented on features/project/source/browse_files.feature:
```diff
@@ -133,6 +143,16 @@ Feature: Project Source Browse Files
+    And I fill the commit message
+    And I click on "Replace file"
+    Then I am redirected to the ".gitignore"
+
```
Does there need to be a step that verifies the contents of the file have actually been replaced?

tsigo replied:
As this is the crux of the feature I agree there should be a test for it.

liyakun replied:
@stanhu @tsigo I also thought about that before, however, I did not figure out how to actually pick one file. Because the feature requires pick one file after click on button "choose file", then the file content can be checked. Do you have any idea? thanks.

tsigo replied:
Looks like there's an [`attach_file`](http://www.rubydoc.info/github/jnicklas/capybara/master/Capybara/Node/Actions#attach_file-instance_method) method in Capybara. Give it the selector for the file field and a path to the new file.

stanhu replied:
Good point. Here's an example:

https://github.com/gitlabhq/gitlabhq/blob/master/features/steps/groups.rb#L131

liyakun replied:
@tsigo @stanhu thanks, as you can see I already use attach_file in the browse_file.rb, however, after attach file to the :file_upload, and then when after click button "Upload file", it is not redirect to the blob path of the uploaded file, I have not idea why it happens here. Except file the commit message, do  I need to do some other actions after attach file and before click the "Upload file" action?

liyakun replied:
@tsigo @stanhu I attach one test image for clear.
![chinesereminder1](https://cloud.githubusercontent.com/assets/7370334/8353498/e169286a-1b3e-11e5-91b8-a89e5002a65b.png)

liyakun replied:
@tsigo @stanhu I let the test to check the content of current page after click "Upload file" button contains "user.feature", then the upload is successful, do you think is this fine?
The previous feature I think is that variable current_path is not the path to current page after upload, so it is not consistent.
![chinesereminder1](https://cloud.githubusercontent.com/assets/7370334/8355111/94561bc8-1b4d-11e5-96b2-ede69b117f67.png)

stanhu commented on features/project/source/browse_files.feature:
```diff
@@ -33,6 +33,16 @@ Feature: Project Source Browse Files
+    When I upload "user.feature"
+    And I fill the commit message
+    And I click on "Upload file"
 
```
Does there need to be a step to verify the file was uploaded?

liyakun commented:
@tsigo do you have other suggestions?

stanhu commented on CHANGELOG:
```diff
@@ -3,6 +3,7 @@ Please view this file on the master branch, on stable branches it's out of date.
 v 7.12.0 (unreleased)
   - Update oauth button logos for Twitter and Google to recommended assets
   - Update browser gem to version 0.8.0 for IE11 support (Stan Hu)
+  - Add "replace" and "upload" functionalities to allows user replace existing file and upload new file into current repository
```
To make this mergeable, please:

1. [rebase master and squash your 54 commits to 1](http://doc.gitlab.com/ee/workflow/gitlab_flow.html#squashing-commits-with-rebase)
2. Move this line from 7.12.0 to 7.13.0
3. Remove the extraneous CHANGELOG entries

tsigo commented on app/views/projects/blob/_actions.html.haml:
```diff
@@ -17,6 +17,6 @@
-  = button_tag class: 'remove-blob btn btn-sm btn-remove',
-      'data-toggle' => 'modal', 'data-target' => '#modal-remove-blob' do
-    Remove
+  .btn-group{"aria-label" => "...", :role => "group"}
```
Please remove the `aria-label` attribute, since "..." won't help screen readers at all.

tsigo commented on app/views/projects/blob/_actions.html.haml:
```diff
@@ -17,6 +17,6 @@
-    Remove
+  .btn-group{"aria-label" => "...", :role => "group"}
+    %button.btn.btn-default{class: 'btn-primary', href:'#modal-replace-blob', 'data-target'=>'#modal-replace-blob', 'data-toggle'=>'modal'} Replace
+    %button.btn.btn-default{class: 'btn-remove', href:'#modal-remove-blob', 'data-target'=>'#modal-remove-blob', 'data-toggle'=>'modal'} Remove
```
Spaces around `=>`

tsigo commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,24 @@
+          From branch
+          %strong= @ref
+      .modal-body
+        = form_tag namespace_project_blob_path(@project.namespace, @project, @id), method: :put, class: 'form-horizontal', :multipart => true  do
```
Use 1.9 Hash syntax here: `multipart: true`

tsigo commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,24 @@
+          %br
+          .form-group
+            .col-sm-offset-2.col-sm-10
+              = file_field_tag :file_upload, :id => "file_upload", :class => "file_up", :required => "required"
```
1.9 Hash syntax.

tsigo commented on app/views/projects/blob/_upload.html.haml:
```diff
@@ -0,0 +1,24 @@
+          From branch
+          %strong= @ref
+      .modal-body
+        = form_tag namespace_project_create_blob_path(@project.namespace, @project, @id), method: :post, class: 'form-horizontal', :multipart => true do
```
1.9 Hash syntax

tsigo commented on app/views/projects/blob/_upload.html.haml:
```diff
@@ -0,0 +1,24 @@
+          %br
+          .form-group
+            .col-sm-offset-2.col-sm-10
+              = file_field_tag :file_upload, :id => "file_upload", :class => "file_up", :required => "required"
```
1.9 Hash syntax

tsigo commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,24 @@
+              = button_tag 'Replace file', class: 'btn btn-small btn-primary btn-remove-file', id: 'replace_btn'
+              = link_to "Cancel", '#', class: "btn btn-cancel", "data-dismiss" => "modal"
+:coffeescript
+  disableButtonIfAnyEmptyField $('#commit_message_replace'), $('.btn-remove-file')
```
These calls have been removed in favor of a new behavior.

See https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/900

tsigo commented on app/views/projects/blob/_upload.html.haml:
```diff
@@ -0,0 +1,24 @@
+              = button_tag 'Upload file', class: 'btn btn-small btn-primary btn-remove-file', id: 'upload_button'
+              = link_to "Cancel", '#', class: "btn btn-cancel", "data-dismiss" => "modal"
+:coffeescript
+  disableButtonIfAnyEmptyField $('#commit_message_replace'), $('.btn-remove-file')
```
These calls have been removed in favor of a new behavior.

See https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/900

tsigo commented on app/views/projects/blob/new.html.haml:
```diff
@@ -1,5 +1,11 @@
-%h3.page-title New file
+%h3.page-title<
+  Create new file or 
+  %a.button{href: '#modal-upload-blob', 'data-target'=>'#modal-upload-blob', 'data-toggle'=>'modal'}> &nbsp;upload existing one
```
Spaces around `=>`

liyakun commented:
@tsigo do you know why after rebase on current master, the test always failed with "Don't know how to build task 'jasmine:ci'" ? thanks.

tsigo commented:
@randx Could you please update Semaphore to run `bundle exec teaspoon` instead of `bundle exec rake jasmine:ci`?

liyakun commented:
@tsigo thanks, as read by here
http://doc.gitlab.com/ce/development/ci_setup.html#build-configuration-on-semaphore-for-testing-the-github.com-repo
The build is set up by the script right, how to set it manually? thanks.

tsigo commented:
@liyakun It's something we have to change on our end, that's why I mentioned randx and not you in that comment. :smile:

randx commented:
@tsigo this should help #9430

liyakun commented:
@tsigo @stanhu the commits are squashed.

stanhu commented on CHANGELOG:
```diff
@@ -6,6 +6,8 @@ v 7.13.0 (unreleased)
   - Add support for unlocking users in admin settings (Stan Hu)
   - Add Irker service configuration options (Stan Hu)
+  - Add "replace" and "upload" functionalities to allow user replace existing file and upload new file into current repository
+  - Add "replace" and "upload" functionalities to allow user replace existing file and upload new file to current repository
```
These two lines seem to be redundant.

stanhu commented on CHANGELOG:
```diff
@@ -140,6 +142,7 @@ v 7.11.0
   - Don't allow a merge request to be merged when its title starts with "WIP".
   - Add a page title to every page.
   - Allow primary email to be set to an email that you've already added.
+  - Get Gitorious importer to work again.
```
This line should not be in the list of changes.

stanhu commented on CHANGELOG:
```diff
@@ -215,6 +218,8 @@ v 7.10.0
   - Don't use HTML ellipsis in EmailsOnPush subject truncated commit message.
   - Set EmailsOnPush reply-to address to committer email when enabled.
+
+v 7.10.0 (unreleased)
```
This should also not be in the list of changes.

stanhu commented on app/views/projects/blob/_remove.html.haml:
```diff
@@ -11,7 +11,7 @@
         = form_tag namespace_project_blob_path(@project.namespace, @project, @id), method: :delete, class: 'form-horizontal js-requires-input' do
           = render 'shared/commit_message_container', params: params,
-                   placeholder: 'Removed this file because...'
+                  placeholder: 'Removed this file because...'
```
Is this change necessary? I think the leading space is desired.

stanhu commented on features/project/source/browse_files.feature:
```diff
@@ -45,7 +56,7 @@ Feature: Project Source Browse Files
     And I should see its new content
 
-  @javascript
+  @javascript @tricky
```
What does the `@tricky` tag do?

stanhu commented on features/steps/project/source/browse_files.rb:
```diff
@@ -106,9 +118,37 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
+  end
+
+  step 'I see the new content of the ".gitignore"' do
+    expect(page).not_to have_content old_gitignore_content
```
Is there a way to make this test positive? For example:

```
expect(page).to have_content new_gitignore_content
```

stanhu replied:
Oh, it looks like there are two steps already:

```
  step 'I should see its content' do
    expect(page).to have_content old_gitignore_content
  end

  step 'I should see its new content' do
    expect(page).to have_content new_gitignore_content
  end
```

liyakun replied:
@stanhu thanks the 'new_gitignore_content' was defined in the project, and I attach a new different file, so I can not use 'new_gitignore_content', but I can use not 'old_gitignore_content'.

liyakun replied:
@stanhu what do you suggest to do here? thanks.

73 commented:
Is that ready to merge now? I appreciate all the effort that is going into this but I'm afraid I lost track within all the comments. What is to be done now? Are there open point still?
Thanks. --73

liyakun commented:
@tsigo @stanhu other things needs to be changed? thanks.

stanhu commented:
The only issue I see is that the test case doesn't seem to be quite right. Please find a way to test that the file contents have been replaced correctly. You could try attaching a fixture and verifying the contents were replaced with that.

liyakun commented:
@stanhu thanks for your advices. I changed the check of "replace", and the scenario is like follow:
1). Replace the file content of ".gitignore" with file content of "LICENSE".
2). Check the replaced content of ".gitignore" contians "Copyright", which is in file "LICENSE".
I attach one debug image of the local feature testing .
![screenshot_2015-07-15-20-30-02 015](https://cloud.githubusercontent.com/assets/7370334/8707557/bafdf39c-2b36-11e5-815c-f308c2bb036c.png)

Do you think this is okay? thanks.

stanhu commented on CHANGELOG:
```diff
@@ -10,6 +10,8 @@ v 7.13.0 (unreleased)
   - Add Irker service configuration options (Stan Hu)
   - Fix order of issues imported from GitHub (Hiroyuki Sato)
+  - Add "replace" and "upload" functionalities to allow user replace existing file and upload new file into current repository
+  - Fix order of issues imported form GitHub (Hiroyuki Sato)
```
Duplicate line

stanhu commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -146,10 +146,19 @@ def editor_variables
+     
+    if params[:file_upload].present?
+      params[:content] = params[:file_upload].read
+      params[:encoding] = 'text'
```
What happens if someone tries to replace a JPG with another one?

liyakun replied:
it will show the download of the file, like follows,
![testjpg](https://cloud.githubusercontent.com/assets/7370334/8708012/af1763ee-2b39-11e5-8ad8-3d384b5ca09c.png)

stanhu replied:
I just tested this out. If you replace an image with another image, the replaced file is corrupted.

stanhu replied:
This goes for any binary file since the forced text encoding corrupts the file.

liyakun replied:
@stanhu I have no this problem locally, that why I removed my previous code for encoding after rebase on the newest master. It still not show this error in my local test(replace any binary file), but if it happens on your side, I will fix it.

liyakun replied:
Could you also send me a screen shot of your error? so I will be more clear about the error, thanks.

stanhu replied:
Here's what happened when I replaced `underscore.png` with another PNG file:

![image](https://cloud.githubusercontent.com/assets/963826/8708657/3061c23e-2af2-11e5-9620-3c500d15c40c.png)

The file sizes also do not match. Correct: 275207 bytes. New: 275205. You can see the diff of the hexdump is also different:

![image](https://cloud.githubusercontent.com/assets/963826/8708689/6981a746-2af2-11e5-9cd8-dd663c850db6.png)

liyakun replied:
@stanhu sorry write back to you late. I tested the file replacement, did you compare the raw file to the original image file?
I replaced "publicmilestone.png" with "gitlab_diagram_overview.png", and compared the raw file of replaced "publicmilestone.png" with "gitlab_diagram_overview.png", they are the same.
![compare_hex](https://cloud.githubusercontent.com/assets/7370334/8849281/c9b840fa-313e-11e5-9c30-998741331d89.png)
However, in the view, the size of the file used to replace is not shown correctly, as it is considered as 'text' encoded, right?
original file in local directory
![original](https://cloud.githubusercontent.com/assets/7370334/8849407/a79060e2-313f-11e5-9698-e6310c08e6f1.png)
replace in view (after being download, this file has the same size and content as original)
![replace](https://cloud.githubusercontent.com/assets/7370334/8849406/a78f603e-313f-11e5-9046-95da24da6f4c.png)

stanhu replied:
First, you should do an `md5` check; the files should be different if they are binary. Can we just disable the replace/upload functionality for binary files?

73 replied:
That would be unfortunate, since there are a lot of binary formats like
doc, ppt, zip, pdf... I think that is the whole point of that feature,
isn't it?

Best regards,
73
On Jul 24, 2015 7:47 PM, "Stan Hu" <notifications@github.com> wrote:

> In app/controllers/projects/blob_controller.rb
> <https://github.com/gitlabhq/gitlabhq/pull/9138#discussion_r35447142>:
>
> >          File.join(@path, File.basename(params[:file_name]))
> >        else
> >          @path
> >        end
> > +
> > +    if params[:file_upload].present?
> > +      params[:content] = params[:file_upload].read
> > +      params[:encoding] = 'text'
>
> First, you should do an md5 check; the files should be different if they
> are binary. Can we just disable the replace/upload functionality for binary
> files?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gitlabhq/gitlabhq/pull/9138/files#r35447142>.
>

stanhu replied:
Perhaps, but then this feature needs to work seamlessly for both text and binary files. 

We also have to consider what happens when the file uploaded is too big. Dropzone used in GitLab prevents users from sending anything over 10 MB, and the Web server also has to be configured appropriately to accept payloads that large.

73 replied:
That is a good point. I should be noted on the modal with the file chooser.
And yes, I totally agree on the seamless integration.
On Jul 24, 2015 7:58 PM, "Stan Hu" <notifications@github.com> wrote:

In app/controllers/projects/blob_controller.rb
<https://github.com/gitlabhq/gitlabhq/pull/9138#discussion_r35448408>:

>          File.join(@path, File.basename(params[:file_name]))
>        else
>          @path
>        end
> +
> +    if params[:file_upload].present?
> +      params[:content] = params[:file_upload].read
> +      params[:encoding] = 'text'

Perhaps, but then this feature needs to work seamlessly for both text and
binary files.

We also have to consider what happens when the file uploaded is too big.
Dropzone used in GitLab prevents users from sending anything over 10 MB,
and the Web server also has to be configured appropriately to accept
payloads that large.

—
Reply to this email directly or view it on GitHub
<https://github.com/gitlabhq/gitlabhq/pull/9138/files#r35448408>.

73 replied:
Do you have a tip regarding implementation?
On Jul 24, 2015 8:04 PM, "Arnold Sykosch" <arnold.sykosch@gmail.com> wrote:

> That is a good point. I should be noted on the modal with the file
> chooser. And yes, I totally agree on the seamless integration.
> On Jul 24, 2015 7:58 PM, "Stan Hu" <notifications@github.com> wrote:
>
> In app/controllers/projects/blob_controller.rb
> <https://github.com/gitlabhq/gitlabhq/pull/9138#discussion_r35448408>:
>
> >          File.join(@path, File.basename(params[:file_name]))
> >        else
> >          @path
> >        end
> > +
> > +    if params[:file_upload].present?
> > +      params[:content] = params[:file_upload].read
> > +      params[:encoding] = 'text'
>
> Perhaps, but then this feature needs to work seamlessly for both text and
> binary files.
>
> We also have to consider what happens when the file uploaded is too big.
> Dropzone used in GitLab prevents users from sending anything over 10 MB,
> and the Web server also has to be configured appropriately to accept
> payloads that large.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gitlabhq/gitlabhq/pull/9138/files#r35448408>.
>
>

liyakun replied:
@stanhu @73, thanks, I also would like to keep the functionality to support binary file upload and replace.  I will try to support both, and tips will be helpful, thanks.

liyakun replied:
@stanhu do you think 
params[:content] = (Base64.encode64(params[:file_upload].read))
params[:encoding] = 'base64'
this encoding is okay for binary file?

liyakun replied:
@stanhu the test of file replacement can be achieved by using md5 check, like:
  step 'I see the new content of the ".gitignore"' do
    Digest::MD5.hexdigest("LICENSE") == Digest::MD5.hexdigest(".gitignore")
    Digest::MD5.hexdigest(".gitignore") != Digest::MD5.hexdigest(old_gitignore_content)
  end
Do you have any suggestions regarding to the uploaded file size ? Thanks.

stanhu replied:
Dropzone currently provides a client-side drag & drop and file size check, which uses the application settings. You may want to look at that first. You may want to look at uploads_controller.rb to see how the server-side check is done for uploads: the upload is stored in a temporary place, verified, and then approved if it passes the file size check.

liyakun replied:
@stanhu do you think the following way to check the file size is okay? The user is redirect to the project repository tree if the file size exceeds 10MB, and the reminding information is also shown.
![file_size](https://cloud.githubusercontent.com/assets/7370334/9134762/b5e0283e-3d07-11e5-8ec2-0e89220bd241.png)
![reminder](https://cloud.githubusercontent.com/assets/7370334/9134763/b5e17a86-3d07-11e5-9d4c-056d3839cae1.png)

stanhu replied:
There are several issues with this approach:

1. There is no client-side check to prevent a big upload from being sent in the first place
2. The 10 MB limit should use the application settings config

Is there a way just to use the existing Dropzone code to make this drag & drop? That way you can take advantage of existing code GitLab is using for this.

liyakun replied:
@stanhu thanks, I have one basic question that I did not find the original file dropzone.js in GitLab, has it been rewriten?

stanhu replied:
No, see:

https://github.com/gitlabhq/gitlabhq/blob/master/app/assets/javascripts/dropzone_input.js.coffee

liyakun replied:
@stanhu , thanks. I am using the 'class: "dropzone"' in file_tag of _upload.html.haml, like:

= form_tag namespace_project_create_blob_path(@project.namespace, @project, @id), method: :post, multipart: true, class: "dropzone" do
          %br
          .form-group
            .col-sm-offset-2.col-sm-10
              = file_field_tag :file_upload, required: true

I use params[:file] in the blob_controller, but it is a nil. I think I misunderstand some things here, could you give me some information? thank you very much.

liyakun replied:
@stanhu I have one question regarding the server side implementation of dropzonejs. I think we need to use the corresponding model to handle the uploaded file from dropzone, and I did not find 'blob model'. I think this is handled by the commit_file in repository model, right?

stanhu replied:
@liyakun I thought it would be easy to get Dropzone working given the [simple tutorial](http://www.dropzonejs.com/#usage) and the existing code base, but if it's proving too difficult we can stick with the Rails upload field. The file size limit may already be restricted by the Web server, so we may just rely on that anyway.

liyakun replied:
@stanhu thanks for your reply. I already got dropzone working with the client side, and I am able to upload the file to repository now. One thing left is that I need to combine the "dropzone file" with the "commit message" in a single form, and I am trying to solve this problem now. It will be good if you have any suggestions, thank you!

liyakun replied:
@stanhu Uploading files through Dropzone is working properly now. I have some questions still,
1. Should we allow multiple files upload?
2. If we allow multiple files upload, then should we redirect user to the tree view rather than blob view after the files are uploaded?  
Thanks.

73 replied:
I guess a single file is okay for a first iteration. Maybe if the feature
is promoted again one may add more functionality?

@stanhu Can this be merged now?

I know, I sound a bit impatient... Sry for that.
--73
On Sep 8, 2015 12:31 PM, "Yakun Li" <notifications@github.com> wrote:

> In app/controllers/projects/blob_controller.rb
> <https://github.com/gitlabhq/gitlabhq/pull/9138#discussion_r38910985>:
>
> >          File.join(@path, File.basename(params[:file_name]))
> >        else
> >          @path
> >        end
> > +
> > +    if params[:file_upload].present?
> > +      params[:content] = params[:file_upload].read
> > +      params[:encoding] = 'text'
>
> @stanhu <https://github.com/stanhu> Uploading files through Dropzone is
> working properly now. I have some questions still,
> 1. Should we allow multiple files upload?
> 2. If we allow multiple files upload, then should we redirect user to the
> tree view rather than blob view after the files are uploaded?
>
> Thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gitlabhq/gitlabhq/pull/9138/files#r38910985>.
>

stanhu replied:
@liyakun Multiple files is probably not necessary at this point. Dropzone can limit the number of files to 1 (see http://stackoverflow.com/questions/18048825/how-to-limit-the-number-of-dropzone-js-files-uploaded/18156897#18156897).

@73 This feature is not ready to be merged, but it will be when it has been properly implemented.

liyakun commented:
@stanhu I changed the code in _upload.html.haml as follow
![_upload_html haml_code](https://cloud.githubusercontent.com/assets/7370334/9630641/f9b6a92e-517b-11e5-8254-3dfec6dbb6e0.png)
I create a new dropzone id, and it support upload file to the repository. There is a problem that the system try to upload file twice, the first time is correct. The second time, the params[:file] is nil in the 'editor_variables' function, and also other attributes. Do you know why this happen? thanks.

liyakun commented:
I think the following information may be helpful. The new file is like follows,
![my_awesome_dropzone](https://cloud.githubusercontent.com/assets/7370334/9631061/a951dce4-517e-11e5-8f32-ab47ea90e0db.png)
The file will be uploaded until the user click the button.

stanhu commented on CHANGELOG:
```diff
@@ -9,7 +9,8 @@ v 7.13.0 (unreleased)
   - Add support for unlocking users in admin settings (Stan Hu)
   - Add Irker service configuration options (Stan Hu)
-  - Fix order of issues imported from GitHub (Hiroyuki Sato)
+  - Add "replace" and "upload" functionalities to allow user replace existing file and upload new file into current repository
```
Please rebase master, move this CHANGELOG entry to 8.0, and ensure the line below is not touched.

liyakun replied:
@stanhu thanks, I am writing the test of file upload with dropzone, but without success yet. Do you have any suggestions?

liyakun committed:
Add "Replace" and "Upload" features

    Refactor upload and replace functionality

    Rename file and move CSS

    Fix typo

    Make dropzone a div

    Remove unnecessary file

    Change color of "upload existing one"

    Add missing changes

stanhu commented:
@liyakun Thanks for the work so far. I have a couple of issues with this:

![image](https://cloud.githubusercontent.com/assets/963826/9796757/a79a9524-57ac-11e5-85eb-80a9ec16ca12.png)

1) The border for the drag and drop upload box looks unclear. Can you add some sort of border to make that clearer?

2) After uploading, I see this alert pop-up 100:

![image](https://cloud.githubusercontent.com/assets/963826/9796885/47230e14-57ad-11e5-8f47-84447397c406.png)

3) What happens if a user attempts to upload the same fileaname again? Is there some sort of warning/message we need to add to show a diff? 

4) In the Replace code, I see this when I attempt to select multiple files:

![image](https://cloud.githubusercontent.com/assets/963826/9796961/a1c4446e-57ad-11e5-8153-8465ecc9c051.png)

Are multiple files upload not implemented here?

5) I also got an `undefined method gsub!` for nil:NilClass when trying to replace a file:

![image](https://cloud.githubusercontent.com/assets/963826/9797033/00ab990a-57ae-11e5-981e-f50e10c43a94.png)

stanhu commented:
Also, does it make more sense to put the filename icons inside the Dropzone box?

stanhu commented on app/assets/javascripts/blob/upload_file.js.coffee:
```diff
@@ -0,0 +1,31 @@
+      return
+    @on 'success', (file, response)->
+      window.location.href = response.filePath
+      alert(gon.max_file_size)
```
Probably can take this out now. :)

stanhu commented on CHANGELOG:
```diff
@@ -165,7 +166,7 @@ v 7.13.0
   - Add support for unlocking users in admin settings (Stan Hu)
   - Add Irker service configuration options (Stan Hu)
-  - Fix order of issues imported from GitHub (Hiroyuki Sato)
+  - Fix order of issues imported form GitHub (Hiroyuki Sato)
```
Please omit this line.

stanhu replied:
@liyakun This change keeps coming back.

liyakun replied:
@stanhu sorry, I will remove it.

liyakun commented:
@stanhu I think the issues 2,4,5 should be solved now. I checked the code in app/service/files/create_service.rb, it has
![samename](https://cloud.githubusercontent.com/assets/7370334/9799009/e6c6d9b2-5804-11e5-946a-1fe8647698e0.png)

stanhu commented on CHANGELOG:
```diff
@@ -172,7 +173,6 @@ v 7.13.0
   - Remove link leading to a 404 error in Deploy Keys page (Stan Hu)
   - Add support for unlocking users in admin settings (Stan Hu)
   - Add Irker service configuration options (Stan Hu)
-  - Fix order of issues imported from GitHub (Hiroyuki Sato)
```
This line should not be deleted.

stanhu replied:
This line should still be there. Please check your diffs carefully!

stanhu commented:
This is starting to look good. I ran into a few other issues:

6) I attempted to upload a new file with an existing filename, but the error message that you mentioned above didn't show up anywhere. The page reloaded as though nothing had happened and loaded the 
existing file.

7) I suggest requiring a commit message in the form. If no commit message is specified, an awkward message `--no-commit-message` shows up:

![image](https://cloud.githubusercontent.com/assets/963826/9799643/67b38c0c-57bd-11e5-864e-e7623ca445b0.png)

8) When you click "Upload" or "Replace", the button should go inactive. Perhaps a spinner icon may be also helpful to show that the request is processing.

9) Similarly, when no files have been added, the "Replace file" should be inactive.

10) When uploading multiple files with relatively long filenames, I see:

![image](https://cloud.githubusercontent.com/assets/963826/9799743/0f38b54c-57be-11e5-80ca-16c3f9870139.png)

stanhu commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -44,11 +50,16 @@ def update
-      flash[:notice] = "Your changes have been successfully committed"
-      redirect_to after_edit_path
+      respond_to do |format|
+        format.html {redirect_to after_edit_path}
```
I'm not sure how this passed Rubocop, but please format these lines as the following:

```ruby
format.html { redirect_to after_edit_path }
```

stanhu commented on features/project/source/browse_files.feature:
```diff
@@ -33,6 +33,13 @@ Feature: Project Source Browse Files
+    Given I click on "new file" link in repo
+    Then I can see new file page
+    And I can see "upload existing one"
+    And I click on "upload existing one"
```
Are you able to verify the upload contents? For example, you may be able to download the file via the raw controller.

liyakun replied:
@stanhu I have not figure out how to drop file to dropzone. I think the above issues 7 and 10 are solved now by disable button and removing additional files.

stanhu replied:
Can you use the Dropzone fallback mode to upload via a standard form and test what you had before?

http://www.dropzonejs.com/

liyakun replied:
@stanhu I test it today without success, the Capybara is not able to find it. I attach one part of the debug source code.
![dropzone](https://cloud.githubusercontent.com/assets/7370334/9802801/432bc8d0-581d-11e5-8117-87e61dde2ee9.png)

stanhu replied:
Yeah, fallback will only happen if JavaScript is disabled. If you comment out the `@javascript` line, do things still work?

stanhu replied:
I think it's fine to have a separate "Select file" button that doesn't rely on JavaScript too.

stanhu replied:
This may be overkill, but here's a good article on how to test Dropzone interaction with Capybara:

http://blog.paulrugelhiatt.com/rails/testing/capybara/dropzonejs/2014/12/29/test-dropzonejs-file-uploads-with-capybara.html

liyakun replied:
@stanhu thanks, remove the "@javascript", Capybara is able to find the file_filed_tag, but the test of file upload is still not success as the params[:file] is nil.

liyakun replied:
@stanhu about the issue 6, I checked that the create_service.rb's checking will work when we are not at the head tree. If I uploaded one file in the sub-folder, then it will work. 

For example,
1). No error prompt when we in head tree 
![tver_head01](https://cloud.githubusercontent.com/assets/7370334/9815767/d5a6c860-5898-11e5-8cfb-f829ba7f0546.png)
![tver_head02](https://cloud.githubusercontent.com/assets/7370334/9815769/d5aa634e-5898-11e5-868f-52e3dc2c9f91.png)
![tver_head03](https://cloud.githubusercontent.com/assets/7370334/9815768/d5a9b96c-5898-11e5-8958-8ff255c0d347.png)

2). error prompt when we are in sub-folder
![tver_sub01](https://cloud.githubusercontent.com/assets/7370334/9815786/f8dff4b4-5898-11e5-9558-5a983de17e7c.png)
![tver_sub02](https://cloud.githubusercontent.com/assets/7370334/9815788/f8e424c6-5898-11e5-88a7-6bd729692853.png)
![tver_sub03](https://cloud.githubusercontent.com/assets/7370334/9815787/f8e2e6ec-5898-11e5-8500-8cda16e8e790.png)

stanhu replied:
It looks like this works for me now for both root directory or subdirectories.

stanhu replied:
The bug is actually in gitlab_git:

https://gitlab.com/gitlab-org/gitlab_git/issues/12

Please revert this change, as we should be stripping leading slashes in that module.

liyakun replied:
Okay, I will revert it, thanks.

Razer6 commented on app/assets/javascripts/blob/upload_file.js.coffee:
```diff
@@ -0,0 +1,35 @@
+      window.location.href = response.filePath
+      return
+    @on 'maxfilesexceeded', (file) ->
+      alert 'Only one file allowed please!'
```
I would rather say "Only one file is allowed" /cc @JobV

JobV replied:
+1

liyakun replied:
@Razer6 @JobV thanks, I will change it.

dosire commented:
Thanks everyone for working on this, would be awesome to have.

Razer6 commented:
One step towards making GitLab usable for non-developers. More manager friendly ;-)

Razer6 commented:
@dosire btw that is something GitHub does not have right now.

liyakun commented:
@dosire @Razer6 thanks ;-)

liyakun commented:
@stanhu I checked the code in create_service.rb, and it seems the file name validation not working for the root folder, but works for the sub-folder. The file in root folder like "LICENSE", the "@file_path" is "/LICENSE", I think this is the reason. 
I changed the code, and it is working for the new feature and online editor at present. I am not sure whether is this appropriate, please let me know if not. Thanks.

dosire commented:
@Razer6 nice!

stanhu commented on app/assets/javascripts/blob/upload_file.js.coffee:
```diff
@@ -0,0 +1,35 @@
+      window.location.href = response.filePath
+      return
+    @on 'maxfilesexceeded', (file) ->
+      alert 'Only one file is allowed'
```
I don't think this alert is necessary.

liyakun replied:
@stanhu I say you already remove this alert in your forked repository https://github.com/stanhu/gitlabhq/commit/ff3ff0c1748b559771c2a1a39bcd229d2fe12725

I also saw your test for dropzone with text file, I also tested it locally and I think it is working properly, thanks. How should we integrate?

stanhu replied:
You can fetch my commit (or download the diff) and apply it to your branch.

stanhu commented on app/assets/javascripts/blob/upload_file.js.coffee:
```diff
@@ -0,0 +1,35 @@
+  parallelUploads: 1,
+  maxFiles: 1,
+  addRemoveLinks: true,
+  previewsContainer: '.dropzone-previews',
```
I suggest adding a border around the file upload box. http://www.dropzonejs.com/ has an example.

liyakun replied:
@stanhu I added a border locally, as the container is also in the form with dropzone class, then it looks like 
![border](https://cloud.githubusercontent.com/assets/7370334/9837647/75c7ad4c-5a46-11e5-89c9-4ca7de0ef180.png)

stanhu replied:
I think the border should be around the area you drag and drop.

stanhu replied:
Can you also make the "click to upload" more obvious? For example, this is another example of how someone uses Dropzone. Notice the border as well:

![image](https://cloud.githubusercontent.com/assets/963826/9837835/5027ba14-5a00-11e5-878b-3f8dfaf66475.png)

stanhu commented:
I'm trying to see if I get your tests to pass. 

Also noting another issue: if Dropzone allows the upload but the Web server rejects it, I see:

![image](https://cloud.githubusercontent.com/assets/963826/9835255/67b31210-5996-11e5-923b-8298399f782b.png)

stanhu commented:
@liyakun I'm pleased to report that I've got basic tests using Dropzone to work. See:

https://github.com/stanhu/gitlabhq/commit/ff3ff0c1748b559771c2a1a39bcd229d2fe12725

Still, I think there are still a few loose ends and UI improvements as mentioned in this issue.

stanhu commented on features/steps/project/source/browse_files.rb:
```diff
@@ -106,9 +119,38 @@ class Spinach::Features::ProjectSourceBrowseFiles < Spinach::FeatureSteps
+  end
+
+  step 'I see the new content of the ".gitignore"' do
+    Digest::MD5.hexdigest("LICENSE") == Digest::MD5.hexdigest(".gitignore")
```
This doesn't do anything. It returns false all the time.

stanhu commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -44,11 +50,16 @@ def update
-      render :edit
+      respond_to do |format|
+        format.html { render :edit }
+        format.json { render json: { message: "failed", filePath: request.referrer } }
```
Is it supposed to be `request.referer` or `request.referrer`?

liyakun replied:
I think it should be request.referrer, which is the previous url.

stanhu replied:
Actually, they are the same:

http://apidock.com/rails/v2.3.2/Rack/Request/referrer

liyakun replied:
;-), thanks.

stanhu commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,27 @@
+            .form-group
+              .col-sm-offset-2.col-sm-10
+                = button_tag 'Replace file', class: 'btn btn-small btn-primary btn-replace-file', id: 'submit-all'
+                = link_to "Cancel", '#', class: "btn btn-cancel", "data-dismiss" => "modal"
```
Is there a reason this class styling is commented out?

stanhu replied:
Is it necessary to have different templates here? _upload.html.haml and and _replace.html.haml seem identical except for a few types.

liyakun replied:
here using the same format as in _remove.html.haml;
_upload.html.haml is using post method, and _replace.html.haml is using put method.

Razer6 replied:
:+1: For just one template, which is reused for both

stanhu commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,27 @@
+                .dropzone-previews
+                  %p.dz-message You can drop a file here or click to upload.
+                  .fallback
+                    = file_field_tag "file", multiple: true, id: "replace_file"
```
Should `multiple` be false?

liyakun replied:
I thought it won't influence here as dropzone will handle file upload.

stanhu replied:
If a user somehow gets to this modal and disables JavaScript, this fallback will be used. Will the controller still work in this mode?

liyakun replied:
you are right, I changed the setting, thanks.

stanhu commented on app/views/projects/blob/new.html.haml:
```diff
@@ -1,5 +1,11 @@
-%h3.page-title New file
+%h3.page-title<
+  Create new file or 
+  %a.button{href: '#modal-upload-blob', 'data-target' => '#modal-upload-blob', 'data-toggle' => 'modal'}> &nbsp;upload existing one
```
With the latest design change, the "upload existing one" isn't a different color:

![image](https://cloud.githubusercontent.com/assets/963826/9837795/a21824ea-59ff-11e5-93d5-322718c601f8.png)

Also, I think it looks better without the space highlighted:

![image](https://cloud.githubusercontent.com/assets/963826/9837803/bb36d84a-59ff-11e5-9ec3-ca809aec00fb.png)

Razer6 replied:
:+1:

liyakun replied:
I changed the space highlight, it looks like this now
![spacehightlight](https://cloud.githubusercontent.com/assets/7370334/9838217/fa8bec6c-5a57-11e5-895e-0df2fc6afcae.png)

stanhu commented on app/assets/javascripts/blob/upload_file.js.coffee:
```diff
@@ -0,0 +1,34 @@
+    @on 'maxfilesexceeded', (file) ->
+      @removeFile file
+      return
+};
```
I think there also needs to be an `on 'error'` handler here to format/limit the message reported back to Dropzone.

stanhu commented on app/services/files/create_service.rb:
```diff
@@ -19,6 +19,10 @@ def validate
       end
 
       unless project.empty_repo?
+        if @file_path.start_with?('/')
```
Might be simpler to do: 

```ruby
# Workaround for https://gitlab.com/gitlab-org/gitlab_git/issues/12
@file_path.slice!(0) if @file_path.start_with?('/')
```

stanhu commented:
@liyakun I spent some time making improvements in the current functionality. Here are a few screenshots:

![image](https://cloud.githubusercontent.com/assets/963826/9845181/f9d9fcac-5a7d-11e5-8c31-6678c8f7d937.png)

And here is what happens when an error occurs:

![image](https://cloud.githubusercontent.com/assets/963826/9845231/568a75f8-5a7e-11e5-9875-c7a4e948faf0.png)

See:

https://github.com/liyakun/gitlabhq/compare/master...stanhu:fix-upload-replace?expand=1

liyakun commented:
@stanhu thank you, this looks much better now, :+1:

liyakun commented:
@stanhu Currently the commit message container is still in the dropzone form, which is looks like 
![whole_form_dropzone](https://cloud.githubusercontent.com/assets/7370334/9849573/5fd8bcbe-5aec-11e5-9255-097ffa5a6f32.png)
The user is able to click under the commit message container to trigger the dropzone input.

I changed the UI now like
![partial_from_dropzone](https://cloud.githubusercontent.com/assets/7370334/9849579/6c55e660-5aec-11e5-8cae-f23ce8bc63b2.png)
The commit message input container is now separated out with the dropzone area.

stanhu commented:
@liyakun Thanks, I improved upon your changes and removed a lot of duplicate code:

https://github.com/liyakun/gitlabhq/compare/master...stanhu:fix-upload-replace?expand=1

I think we should reuse _upload.html.haml and _replace.html.haml a single file. Too much duplicate work there.

stanhu commented:
@liyakun I think we're close here! Please test out the latest changes and make sure it all works.

If tests pass, then let's squash this into one commit and clean up the commit messages. Currently https://github.com/liyakun/gitlabhq/commit/57a782b33c3d27691c931d5ee7349bc7cb40e733 is too long and unnecessary.

stanhu commented on app/controllers/projects/blob_controller.rb:
```diff
@@ -44,11 +50,16 @@ def update
     result = Files::UpdateService.new(@project, current_user, @commit_params).execute
 
     if result[:status] == :success
-      flash[:notice] = "Your changes have been successfully committed"
```
Should this `flash[:notice]` go into the HTML render below?

stanhu replied:
Or should it remain as above?

liyakun replied:
I think it should remain at above, so even redirect_to or render not working, the user still know the result of post or put.

stanhu replied:
Sounds good.

stanhu replied:
Please remember to revert this then.

stanhu commented on app/views/projects/blob/new.html.haml:
```diff
@@ -1,5 +1,11 @@
-%h3.page-title New file
+%h3.page-title<
+  Create new file or&nbsp;
+  %a{href: '#modal-upload-blob', 'data-target' => '#modal-upload-blob', 'data-toggle' => 'modal'}>upload existing one
```
"upload existing one" probably needs a new color:

![image](https://cloud.githubusercontent.com/assets/963826/9855084/2122fd70-5ac1-11e5-906c-04d3e8ae00f8.png)

Or perhaps some icon to indicate? It's not obvious that this is a link.

liyakun replied:
I changed the color to blue, it seems more obvious.
![color](https://cloud.githubusercontent.com/assets/7370334/9855867/4c628424-5b11-11e5-8f8f-5f04377d5971.png)

stanhu commented on app/views/projects/blob/_replace.html.haml:
```diff
@@ -0,0 +1,28 @@
+    .modal-content
+      .modal-header
+        %a.close{href: "#", "data-dismiss" => "modal"} ×
+        %h3.page-title Replace
```
I noticed you removed the blob name from the title. Any reason?

liyakun replied:
sorry, It should have, I think I removed during change between upload and replace.

stanhu commented on app/assets/javascripts/blob/blob_file_dropzone.js.coffee:
```diff
@@ -0,0 +1,52 @@
+        return
+    )
+
+    submitButton = document.querySelector('#submit-all')
```
I think this can be changed to:

```coffeescript
submitButton = form.find('#submit-all')[0]
```

liyakun replied:
yes, changed, thanks.

stanhu commented:
@liyakun Great, thanks! Can you squash this feature into one commit and clean up the final commit message?

Razer6 commented:
Thanks @liyakun and thanks @stanhu. I would :heart: to see that in 8.0. My manager :heart: you either.

liyakun commented:
@stanhu I checked the new feature with invalidate file name upload (@^$6.txt) before, and it is working correctly. 
![before](https://cloud.githubusercontent.com/assets/7370334/9858985/2f26dc8a-5b24-11e5-855a-58baf1c6c6b1.png)
However, after I get the newest version of git-development-kit, and with newest merging, it is not working correctly, like
![after](https://cloud.githubusercontent.com/assets/7370334/9859029/81bec0b6-5b24-11e5-844a-8ed7254ab21b.png)

It seems the request.referer is not working properly, I am trying to find solutions now.

liyakun commented:
@Razer6 thank you. ;-)

stanhu commented:
@liyakun Yeah, for some reason the `Referer` field shows up on my browser as the following:

![image](https://cloud.githubusercontent.com/assets/963826/9859547/0bb735b2-5adc-11e5-9116-ab3f40e6fa36.png)

Maybe because when Dropzone makes the POST request, it doesn't know?

stanhu commented:
Instead of relying on the referer, could we just redirect pack to the original path?

stanhu commented:
Using `after_edit_path` seems to do the trick.

liyakun commented:
@stanhu It seems after_edit_path direct to the tree view.

stanhu commented:
Right. I suppose you can create a path back to the new edit page again.

liyakun commented:
@stanhu Now, user will be directed to new edit page if upload failed.

stanhu commented:
@liyakun, awesome. Let's get this change in--just need the commits squashed and cleaned up.

stanhu commented:
@liyakun I did a diff with the last working branch. It looks like your master is missing a few changes:

```diff
--- a/app/views/projects/blob/show.html.haml
+++ b/app/views/projects/blob/show.html.haml
@@ -10,4 +10,3 @@
 
 - if allowed_tree_edit?
   = render 'projects/blob/remove'
-  = render 'projects/blob/replace'
diff --git a/config/routes.rb b/config/routes.rb
index 54e109f..fad6b3c 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -358,16 +358,6 @@ Gitlab::Application.routes.draw do
             to: 'blob#destroy',
             constraints: { id: /.+/, format: false }
           )
-          put(
-            '/blob/*id',
-            to: 'blob#update',
-            constraints: { id: /.+/, format: false }
-          )
-          post(
-            '/blob/*id',
-            to: 'blob#create',
-            constraints: { id: /.+/, format: false }
-          )
         end
```

stanhu commented:
Great, thanks. Will need to squash that last commit too. Can you please clean up your commit message so that it has a proper description on the first line? You can remove all the extra cruft with the filenames in the message. Right now I see this:

![image](https://cloud.githubusercontent.com/assets/963826/9864193/e20e69e0-5afb-11e5-9f14-ecfa143e3ae7.png)

All these "modified" lines can also be removed in the commit log. For example:

![image](https://cloud.githubusercontent.com/assets/963826/9864212/1bb845f8-5afc-11e5-9df6-9241a6483b33.png)

liyakun commented:
@stanhu thanks, I will squash and change description after the checking pass.

stanhu commented:
@liyakun It "only" took 5 months to get this in, but it's finally done. :) Thank you for your work.

liyakun commented:
@stanhu thank you :)

dosire commented:
Thanks @liyakun and @stanhu

jvanbaarsen commented:
@liyakun & @stanhu thanks for all the hard work! :heart:

JobV commented:
Wow, great work and amazing seeing you all kept focus and pulled this on through! 

:beers: / :wine: / :tropical_drink:  for @liyakun and @stanhu!

liyakun commented:
@dosire @jvanbaarsen @JobV thank you :-)

randx commented on app/assets/stylesheets/pages/tree.scss:
```diff
@@ -116,3 +116,15 @@
+
+.upload-link {
+  font-weight: normal;
+  color: #0000EE;
```
Thats out of style. I will fix it

stanhu replied:
Thanks. Gathering UI feedback here: https://gitlab.com/gitlab-org/gitlab-ce/issues/2520

tsigo commented:
@liyakun We want to be sure to give you proper recognition in the upcoming 8.0 blog post when we mention this feature. Is your surname Li or Yakun?

liyakun commented:
@tsigo thanks, my surname is Li.

tsigo commented:
@liyakun Great, I was right. :smile:
